Library/Homebrew/dev-cmd/audit.rb

```diff
@@ -109,7 +110,11 @@ def audit
 
     # Check style in a single batch run up front for performance
     style_results = Style.check_style_json(style_files, options) if style_files
-
+    # load licenses
+    full_path = File.join(File.dirname(__FILE__), "../data/spdx.json")
+    spdx_ids = File.open(full_path, "r") do |f|
+      JSON.parse(f.read)
```

Comment:
Would we need the full JSON file or would a simple e.g. Ruby array be sufficient for just the SPDX IDs?

Reply:
I think the JSON is small enough (66kb) such that loading it will not constitute a huge issue for us. 
Some of the fields such as `name` `url` and `osiApproved` may also be useful in the future for users to check via `brew info` - a very possible extension in my opinion. 

I actually compared this to its even [fuller alternative](https://github.com/sindresorhus/spdx-license-list/blob/master/spdx-full.json), but felt that the text may be an overkill.

Reply:
Is there a way to leave this data to someone else to maintain? https://github.com/github/choosealicense.com for example

Reply:
How about this (repository)[https://github.com/sindresorhus/spdx-license-list/blob/master/spdx.json
]? The file was last updated in April with 6 contributors. The repository you referenced might required some parsing because each license is stored as an individual file in  the`_license` folder.

Raw Version: https://raw.githubusercontent.com/sindresorhus/spdx-license-list/master/spdx.json

I am thinking we can do a `wget` and cache each time someone runs `brew update`. I wonder if that's an acceptable approach as compared to cloning. the github repository and pulling each time.

Reply:
What about pulling licensee in as a dev dependency? It has an API to get all licences and uses the github list of licenses. 

(I'd trust Microsoft more to support their list long term than I do a random other person)

Reply:
`licensee` is overkill for our purposes and has some native gem build requirements that are both error-prone and time consuming. If we really, really don't want this in the repository then we should hit the SPDX website in `audit` and require `--online`.

Reply:
Another middle ground would be storing this in the repo but having some CI that fails if the file differs (so we know to update it).

Reply:
> Another middle ground would be storing this in the repo but having some CI that fails if the file differs (so we know to update it).

I think this is a good idea - should the CI work be in the scope of this PR?

Reply:
I suppose the ground truth can be referenced against [this version](https://raw.githubusercontent.com/sindresorhus/spdx-license-list/master/spdx.json)
 
I feel that if were to require --onlline even for the most basic level of check (checking if it is a standard spdx license), it limits the usefulness of the audit.

Reply:
> I think this is a good idea - should the CI work be in the scope of this PR?

Yeh, I think so!

Reply:
Hi @MikeMcQuaid and @SMillerDev, 

So[this file](https://github.com/spdx/license-list-data/blob/master/json/licenses.json) seems like it's well maintained because it's auto generated by a [travis ci process ](https://github.com/spdx/license-list-XML/blob/master/.travis.yml)that triggers each time the [license data](https://github.com/spdx/license-list-XML) is updated. Above all, it's supported by the SPDX org. 

What's nice is that it also has a `licenseListVersion` field so that comparison time is constant. In terms of maintenance support, I would say it is pretty reliable! 

However, I looked through and I find that the format the data is saved is not very desirable. Preferably, I would like a json structure that is similar to a hash so I can do O(1) look up when checking validity. This will have performance implications when someone runs `brew audit --new-formula`. 

We thought about it and felt that we could run (1) a CI task to check if the version differs, and if it does differ, the CI will fail, as suggested by Mike. As well as (2) a second task which generates a json file from the old schema to a new schema that ideally looks like [this](https://raw.githubusercontent.com/sindresorhus/spdx-license-list/master/spdx.json) and store it into the repository. We can delete the old file but keep the `licenseListVersion` info in the generated file. 

For (1), to get the developer to update the license, we can create a new command. like `brew license` which automatically pulls from the spdx-maintained repo, similar to how `brew man` works.

What do you think? 

P.S. I realised, from the need to commit the modified `man pages` myself, why can't the CI automatically do that for us? For (1), is it possible that the CI, instead of telling us something has failed, just resolve the failure for us by pulling the new version? Similar to why we flag an error with `brew style` instead of helping the user run `brew style --fixed`. Is there a reason, technical or otherwise, we flag failures instead of fix them in the CI process?

Reply:
> However, I looked through and I find that the format the data is saved is not very desirable. Preferably, I would like a json structure that is similar to a hash so I can do O(1) look up when checking validity. This will have performance implications when someone runs `brew audit --new-formula`.

The O(1) vs. O(n) lookup time only really matters for (very) large values of `n` and in places where performance really matters. I'm not convinced there will be a noticeable slowdown for users doing it as-is.

> For (1), to get the developer to update the license, we can create a new command. like `brew license` which automatically pulls from the spdx-maintained repo, similar to how `brew man` works.

`brew update-spdx-licenses` or something would work.

> I realised, from the need to commit the modified `man pages` myself, why can't the CI automatically do that for us?

It could (in most cases if users allow commit access to their forks) but it's sometimes changed accidentally so I think the user input here is useful and necessary (particularly for those who are less familiar with Git and others pushing to their branch).